applications/yatima.md

```diff
@@ -0,0 +1,439 @@
+  implemented in Rust featuring substructural types, metaprogramming, and
+  content addressing via IPFS.
+
+* Relevance to Substrate: The language core, including parsing, elaboration,
+  serialization/deserialization, typechecking, evaluation, etc. can compile to
+  Substrate compatible WASM (`no_std` via Substrate's [`sp_std` standard library](https://docs.rs/sp-std/3.0.0/sp_std/index.html)), and thus can be
+  utilized within a Substrate Pallet.
+
+* Our interest: In order for decentralized technologies to gain widespread
+  adoption, users have to trust that the infrastructure they are using is safe,
+  and that they are not just replacing counterparty risk with technical risk. We
+  believe that functional programming and formal proofs offer a possible path
+  towards increasing the safety and reliability of all software, not just in
+  blockchain.
+
+### Project Details
+
+Many blockchain projects are exploring strongly-typed functional programming as
+a paradigm for writing smart contracts and DApps. Examples include [Cardano's
+Plutus](https://developers.cardano.org/en/programming-languages/plutus/overview/), [Tezos' Michelson](https://wiki.tezosagora.org/learn/smartcontracts/michelson), [Kadena's Pact](https://github.com/kadena-io/pact) and [Digital Asset's DAML](https://daml.com/). The type safety and
+composability of functional programs make them an exciting option for
+writing high-assurance smart contracts.
+
+Yatima aims to help develop the functional programming ecosystem on Substrate
+and the Polkadot ecosystem. We have developed a graph-based runtime for the Yatima language
+with fully lazy evaluation, sharing, and whole program optimization. This runtime could also 
+be targeted by other functional languages. The Yatima runtime is
+similar in many respects to Haskell Core, or the [GRIN
+IR](https://grin-compiler.github.io/) but is implemented in deterministic
+`no_std` Rust, which allows it to run virtually anywhere, including the web,
+WASM blockchains, or even embedded systems (in theory).
+
+Yatima also implements a powerful dependent and substructural type-system,
+similar to Idris 2. This enables formal verification and theorem proving
+(although crucial features for practical proving such as inference and tactics
+are not yet implemented). Proofs allow for many useful optimizations in the area
+of smart contracts, such as static detection of arithmetic overflow or division
+by zero errors, or tracking of resources and permissions via an effect system.
+Furthermore, since Yatima can be implemented in itself, there is a possiblity of
+having on-chain trustless proof verification.
+
+Lastly, Yatima implements content-addressing, allowing all programs and packages
+to be serialized into IPLD objects and shared over the peer-to-peer IPFS network
+(or via libp2p).
+
+#### Deliverables
+
+- A `sp_ipld` crate for interacting with IPFS CID, Multihash and IPLD based on
```

Comment:
can you integrate concrete feature specifications for these in the description of milestone 1 below?

Reply:
Sure, what concrete feature specifications should we add? Milestone 1 lists:

"serializing and deserializing IPLD objects and computing IPFS CIDs"

Which is the essential functionality, but we can also add a description of the dependencies:

- bytecursor (no_std replacement for Rust Read/Write traits, WIP repo is here: https://github.com/yatima-inc/bytecursor)
- sp-cid (no_std port of https://github.com/multiformats/rust-cid, WIP repo is here: https://github.com/yatima-inc/sp-cid
- sp-multihash (no_std port of https://github.com/multiformats/rust-multihash, WIP repo is here: https://github.com/yatima-inc/sp-multihash)

Also we should clarify that we were only planning on having `sp-ipld` support dag-cbor, but I believe we could prettily add dag-json support (and a json textual format for both) if this is desirable.

Reply:
I've gone ahead and added these to Milestone 1, since the dependencies are implicit anyway, and `dag-json` is a great additional feature to have.